crates/runtime/src/memory.rs

```diff
@@ -374,7 +450,138 @@ impl RuntimeLinearMemory for ExternalMemory {
         }
     }
 
-    #[cfg(feature = "pooling-allocator")]
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+}
+
+/// For shared memory (and only for shared memory), this lock-version restricts
+/// access when growing the memory or checking its size. This is to conform with
+/// the [thread proposal]: "When `IsSharedArrayBuffer(...)` is true, the return
+/// value should be the result of an atomic read-modify-write of the new size to
+/// the internal `length` slot."
+///
+/// [thread proposal]:
+///     https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#webassemblymemoryprototypegrow
+#[derive(Clone)]
+pub struct SharedMemory(Arc<RwLock<SharedMemoryInner>>);
+impl SharedMemory {
+    /// Construct a new [`SharedMemory`].
+    pub fn new(plan: MemoryPlan) -> Result<Self> {
+        assert!(plan.memory.shared);
+        assert!(matches!(plan.style, MemoryStyle::Static { .. }));
```

Comment:
Can these asserts be deferred to `SharedMemory::wrap` and can this method call `SharedMemory::wrap`? Otherwise `Memory::new_dynamic` which I believe is called for locally defined memories isn't asserting these invariants (when it needs to be doing so)